diff --git a/node_modules/testcafe/lib/cli/cli.js b/node_modules/testcafe/lib/cli/cli.js
index 14b9cc6..88633fd 100644
--- a/node_modules/testcafe/lib/cli/cli.js
+++ b/node_modules/testcafe/lib/cli/cli.js
@@ -79,7 +79,35 @@ async function runTests(argParser) {
         .browsers(browsers)
         .reporter(argParser.opts.reporter)
         .concurrency(argParser.opts.concurrency)
-        .filter(argParser.opts.filter)
+        .filter((testName, fixtureName, fixturePath, testMeta, fixtureMeta) => {
+            if (opts.test && testName !== opts.test)
+                return false;
+            if (opts.testGrep && !opts.testGrep.test(testName))
+                return false;
+            if (opts.fixture && fixtureName !== opts.fixture)
+                return false;
+            if (opts.fixtureGrep && !opts.fixtureGrep.test(fixtureName))
+                return false;
+
+            // FIX 1: I fix the issue with tests metas inherit
+            // @see https://github.com/DevExpress/testcafe/issues/4008
+            if (opts.fixtureMeta)
+              throw new Error('Please use --test-meta instead.');
+            if (opts.testMeta) {
+              const _ = require('lodash');
+              const allMeta = _.assignIn({}, fixtureMeta, testMeta);
+              return _.isMatchWith(allMeta, opts.testMeta, (allMetaValue, testMetaValue) => {
+                // FIX 2: I fix the issue with metas arrays: be able to launch tests with meta values in an array
+                if (_.isArray(allMetaValue)) {
+                  return allMetaValue.includes(testMetaValue);
+                } else {
+                  return allMetaValue === testMetaValue;
+                }
+              });
+            }
+
+            return true;
+        })
         .video(opts.video, opts.videoOptions, opts.videoEncodingOptions)
         .screenshots(opts.screenshots)
         .startApp(opts.app, opts.appInitDelay)
